### PR TITLE
Correctness fix PP+ZeRO for gradient accumulation

### PR DESCRIPTION
**Pull request by jeffra**

### CODE REVIEW

**Review by samyam — APPROVED** [review]

File: deepspeed/runtime/engine.py [hunk 1]
```diff
@@ -1145,6 +1145,10 @@ def forward(self, *inputs, **kwargs):
         return loss
 
     def allreduce_gradients(self, bucket_size=MEMORY_OPT_ALLREDUCE_SIZE):
+        # Pass (PP) gas boundary flag to optimizer (required for zero)
+        self.optimizer.is_gradient_accumulation_boundary = self.is_gradient_accumulation_boundary(
```
Comment (tjruwase) [inline deepspeed/runtime/engine.py:1149]: Could self.optimizer be an fp32 optimizer, our fp16 optimizer, or client optimizer? Is it safe to assume `self.optimizer.is_gradient_accumulation_boundary`?

**Comment by MichaelEk** [comment]
This PR fixes this issue: https://github.com/microsoft/DeepSpeed/issues/1217

**Comment by stas00** [comment]
@jeffra, could you please replay to @big-science? Thank you!

**Comment by jeffra** [comment]
I believe this is already replayed on big-science branch: https://github.com/microsoft/DeepSpeed/commit/f93e22b38c3a90d1f9980ea99b3c37a1d5a4e4f9

I think this commit was mixed with a few other items as well though.

**Comment by stas00** [comment]
Indeed. Thank you for finding it! I was just comparing the titles of the commits.

**Pull request merged**